rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/geo/TbGpsMultiGeofencingActionNodeConfiguration.java

```diff
@@ -0,0 +1,55 @@
+/**
+ * Copyright © 2016-2024 The Thingsboard Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.thingsboard.rule.engine.geo;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import lombok.Data;
+import org.thingsboard.server.common.data.EntityType;
+import org.thingsboard.server.common.data.id.DeviceId;
+import org.thingsboard.server.common.data.relation.EntityRelationsQuery;
+import org.thingsboard.server.common.data.relation.EntitySearchDirection;
+import org.thingsboard.server.common.data.relation.RelationEntityTypeFilter;
+import org.thingsboard.server.common.data.relation.RelationsSearchParameters;
+
+import java.util.List;
+import java.util.UUID;
+
+@Data
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class TbGpsMultiGeofencingActionNodeConfiguration extends TbGpsGeofencingFilterNodeConfiguration {
+
+    private EntityRelationsQuery entityRelationsQuery;
+    private EntitySearchDirection relationField;
+    private String perimeterAttributeKey;
+    private Long insideDurationMs;
+    private Long outsideDurationMs;
+
+    @Override
+    public TbGpsMultiGeofencingActionNodeConfiguration defaultConfiguration() {
+        TbGpsMultiGeofencingActionNodeConfiguration configuration = new TbGpsMultiGeofencingActionNodeConfiguration();
+        configuration.setPerimeterAttributeKey("geofences");
+        RelationsSearchParameters relationsSearchParameters = new RelationsSearchParameters(new DeviceId(UUID.randomUUID()), EntitySearchDirection.FROM, 1, true);
+        EntityRelationsQuery entityRelationsQuery = new EntityRelationsQuery();
+        entityRelationsQuery.setParameters(relationsSearchParameters);
+        RelationEntityTypeFilter filter = new RelationEntityTypeFilter("DeviceToZone", List.of(EntityType.ASSET));
+        entityRelationsQuery.setFilters(List.of(filter));
+        configuration.setEntityRelationsQuery(entityRelationsQuery);
```

Comment:
It will be nice to re-use RelationsQuery and EntitiesRelatedEntityIdAsyncLoader. Currently, the loader can return only one entity, however as per my understanding you need to retrieve all the entities based on the relation query, so you can add another method to the loader that will return a list of entities.

Another question:

Search by a relation query is not cacheable and as a result is less efficient than cacheable methods:

    ListenableFuture<List<EntityRelation>> findByFromAndTypeAsync(TenantId tenantId, EntityId from, String relationType, RelationTypeGroup typeGroup);
    ListenableFuture<List<EntityRelation>> findByToAndTypeAsync(TenantId tenantId, EntityId to, String relationType, RelationTypeGroup typeGroup);
    
I'd recommend considering if we can use cacheable methods instead of a relation query. Does it fit the use case?